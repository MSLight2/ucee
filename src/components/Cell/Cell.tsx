import React, { useCallback } from 'react';
import classnames from 'classnames';
import Icon from '../Icon';
import { getIsImage, transformSize } from '../../utils';
import './index.scss';

export interface CellProps {
  icon?: React.ReactNode
  title?: React.ReactNode
  titleWidth?: string | number
  value?: string
  label?: string
  size?: 'default' | 'large'
  border?: boolean
  center?: boolean
  url?: string
  isLink?: boolean
  arrowDirection?: 'left' | 'up' | 'down'
  onClick?: React.MouseEventHandler<HTMLElement>
  className?: string
  titleClass?: string
  labelClass?: string
  valueClass?: string
}

const Cell: React.FC<CellProps> = (props) => {
  const {
    icon,
    title,
    titleWidth,
    value,
    label,
    size,
    border = true,
    center,
    url,
    isLink,
    arrowDirection = 'left',
    onClick,
    className,
    titleClass,
    labelClass,
    valueClass
  } = props;
  const baseClass = 'ucee-cell'
  const cellClassNames = classnames(
    baseClass,
    {
      [`${baseClass}--large`]: size === 'large',
      [`${baseClass}--borderless`]: !border,
      [`${baseClass}--center`]: center,
      [`${baseClass}--active`]: isLink || url,
    },
    className
  );
  const titleClassNames = classnames(
    `${baseClass}__title`,
    titleClass
  );
  const labelClassNames = classnames(
    `${baseClass}__label`,
    labelClass
  );
  const valueClassNames = classnames(
    `${baseClass}__value`,
    valueClass
  );

  const TitleTag = (title && typeof title === 'string') ? 'span' : 'div';
  const MainTag = url ? 'a' : 'div'
  const styleTitleWidth = transformSize(titleWidth);
  const styleProp: any = {}
  if (titleWidth) {
    styleProp.style = { flexGrow: 0, flexBasis: styleTitleWidth };
  }

  const hrefProps: React.AnchorHTMLAttributes<HTMLElement> = {};
  if (url) {
    hrefProps.href = url
  }

  const RenderIcon = useCallback(() => {
    return (
      <>
        {
          getIsImage(icon as string)
            ? <img src={icon as string} className={`${baseClass}__img`} alt="" />
            : icon
          }
      </>
    )
  }, [icon]);

  const ArrowIcon = useCallback(() => {
    let iocnName = 'enter';
    if (arrowDirection === 'down') {
      iocnName = 'unfold';
    }
    if (arrowDirection === 'up') {
      iocnName = 'packup';
    }
    return isLink
      ? <Icon
          name={iocnName}
          size={19}
          className={isLink && `${baseClass}--icon-right`}
        />
      : null
  }, [isLink, arrowDirection]);

  return (
    <MainTag
      className={cellClassNames}
      onClick={(e: any) => onClick?.(e)}
      {...hrefProps}
    >
      <div
        className={titleClassNames}
        {...styleProp}
      >
        <RenderIcon />
        <TitleTag>{title}</TitleTag>
        {label && <div className={labelClassNames}>{label}</div>}
      </div>
      {value && <div className={valueClassNames}>{value}</div>}
      <ArrowIcon />
    </MainTag>
  )
}

export default Cell;