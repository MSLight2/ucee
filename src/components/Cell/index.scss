@import '../../styles/var';
@import '../../styles/mixin';

.ucee-cell{
  position: relative;
  display: flex;
  width: 100%;
  padding: $cell-top-padding $cell-up-padding;
  box-sizing: border-box;
  line-height: $cell-line-height;
  font-size: $cell-font-size;
  background-color: $white;
  color: $text-color;
  text-decoration: none;
  overflow: hidden;

  &::after {
    @include hairline('bottom', $gray-3, 16px, 16px);
  }
  &:last-child::after,
  &--borderless::after{
    display: none;
  }
  &__title,
  &__value{
    flex: 1;
  }
  &__title{
    .ucee-icon{
      vertical-align: middle;
      margin-right: 4px;
    }
  }
  &__value{
    position: relative;
    color: $cell-value-color;
    text-align: right;
    vertical-align: middle;
    overflow: hidden;
  }
  &__label{
    margin-top: $cell-label-margin-top;
    color: $cell-value-color;
    line-height: $cell-label-line-height;
    font-size: $cell-label-font-size;
  }
  &__img{
    display: inline-block;
    width: $cell-img-size;
    height: $cell-img-size;
    margin-right: 4px;
    vertical-align: middle;
  }
  &--large{
    padding-top: $cell-large-top-padding;
    padding-bottom: $cell-large-top-padding;
    .ucee-cell__title{
      font-size: $cell-font-size + 2;
    }
    .ucee-cell__label{
      font-size: $cell-label-font-size + 2;
    }
  }
  &--icon-right{
    color: $cell-icon-color;
    line-height: $cell-line-height;
  }
  &--center{
    align-items: center;
  }
  &--active{
    &:active{
      background: $gray-2;
    }
  }
}