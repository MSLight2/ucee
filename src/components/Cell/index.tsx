import Cell from './Cell';

export type { CellProps } from './Cell';
export default Cell;