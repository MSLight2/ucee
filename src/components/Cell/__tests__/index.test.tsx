import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Cell from '..';

describe('Cell', () => {
  test('should render custom title width', () => {
    const tree = renderer.create(
      <Cell title="单元格" value="内容" titleWidth={300} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('should render down or up arrow icon', () => {
    const tree = renderer.create(
      <>
        <Cell title="单元格" value="内容" arrowDirection="down" />
        <Cell title="单元格" value="内容" arrowDirection="up" />
      </>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('should render link tag', () => {
    const { container } = render(
      <Cell url="about:blank" isLink />
    );

    expect(container.querySelector('.ucee-cell')?.tagName.toLowerCase()).toEqual('a');
  });

  test('should render lable', () => {
    const { container } = render(
      <Cell title="单元格" label='描述信息' />
    );
    const lable = container.querySelector('.ucee-cell__title')?.lastChild;
    
    expect(lable).toHaveClass('ucee-cell__label');
    expect(lable).toHaveTextContent('描述信息');
  });

  test('cell click should trigger event', () => {
    const handlerClick = jest.fn();
    const { container } = render(
      <Cell data-testid="mockClickCell" onClick={handlerClick} />
    );

    fireEvent.click(container.querySelector('.ucee-cell') as Element);
    expect(handlerClick).toHaveBeenCalled();
  });
});