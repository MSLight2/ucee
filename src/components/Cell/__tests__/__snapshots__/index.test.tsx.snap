// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell should render custom title width 1`] = `
<div
  className="ucee-cell"
  onClick={[Function]}
>
  <div
    className="ucee-cell__title"
    style={
      Object {
        "flexBasis": "300px",
        "flexGrow": 0,
      }
    }
  >
    <span>
      单元格
    </span>
  </div>
  <div
    className="ucee-cell__value"
  >
    内容
  </div>
</div>
`;

exports[`Cell should render down or up arrow icon 1`] = `
Array [
  <div
    className="ucee-cell"
    onClick={[Function]}
  >
    <div
      className="ucee-cell__title"
    >
      <span>
        单元格
      </span>
    </div>
    <div
      className="ucee-cell__value"
    >
      内容
    </div>
  </div>,
  <div
    className="ucee-cell"
    onClick={[Function]}
  >
    <div
      className="ucee-cell__title"
    >
      <span>
        单元格
      </span>
    </div>
    <div
      className="ucee-cell__value"
    >
      内容
    </div>
  </div>,
]
`;
